### PR TITLE
Move LibGit2 to stdlib

### PR DESCRIPTION
**Pull request by KristofferC**
Made on top of #25705.

I have a few comments inline.

There was some problems with the documentation generation locally, so we will see what happens here.

### CODE REVIEW

File: base/methodshow.jl [hunk 1]
```diff
@@ -212,6 +212,8 @@ function url(m::Method)
             return "https://github.com/JuliaLang/julia/tree/$(Base.GIT_VERSION_INFO.commit)/base/$file#L$line"
         end
     else
+        return fileurl(file)
+        #=
```
Comment (KristofferC) [inline base/methodshow.jl]: Need to figure out what to do with this one `url(::Method)`...

File: test/docs.jl [hunk 2]
```diff
@@ -887,30 +887,30 @@ let x = Binding(Base, Symbol("@time"))
...
+    @test @var(Base.Iterators.@time) == x
 end
 
 let x = Binding(Iterators, :enumerate)
     @test defined(x) == true
     @test @var(enumerate) == x
     @test @var(Base.enumerate) == x
     @test @var(Iterators.enumerate) == x
-    @test @var(Base.Pkg.Dir.enumerate) == x
+    # @test @var(Base.Pkg.Dir.enumerate) == x
```
Comment (KristofferC) [inline test/docs.jl]: Not sure exactly what this is testing, perhaps just a deep nesting.

File: base/deprecated.jl [hunk 3]
```diff
@@ -243,19 +243,6 @@ DEPRECATED: use @__MODULE__ instead
 end
 export current_module
 
-# PR #22062
```
Comment (vchuravy) [inline base/deprecated.jl]: These deprecations were made during the 0.7 cycle so they need to use `@deprecate_moved` or `@deprecate_stdlib` after #25692

**Comment by omus** [comment]
Is it possible to move out the libgit2 C library with this? It would be nice to be able to update the revision of the shared library during 1.0.

File: doc/src/manual/packages.md [hunk 4]
```diff
@@ -3,8 +3,8 @@
 Julia has a built-in package manager for installing add-on functionality written in Julia. It
 can also install external libraries using your operating system's standard system for doing so,
 or by compiling from source. The list of registered Julia packages can be found at [http://pkg.julialang.org](http://pkg.julialang.org).
-All package manager commands are found in the `Pkg` module, included in Julia's `Base`
-install.
+All package manager commands are found in the `Pkg` standard library, it becomes available after using
```
Comment (waldyrious) [inline doc/src/manual/packages.md]: Suggestion: "it becomes" --> "which becomes", to avoid a [comma splice](https://en.wikipedia.org/wiki/Comma_splice).

**Comment by StefanKarpinski** [comment]
> Is it possible to move out the libgit2 C library with this? It would be nice to be able to update the revision of the shared library during 1.0.

1.0 API stability *definitely* does not include never changing versions of libraries that we depend on.

**Comment by KristofferC** [comment]
So this causes the documentation warnings:

```
 !! Invalid local link: unresolved path
    '../../devdocs/libgit2/#Base.LibGit2.GitRevWalker' in devdocs/libgit2.md
 !! Invalid local link: unresolved path
    '../../devdocs/libgit2/#Base.LibGit2.GitHash' in devdocs/libgit2.md
 !! Invalid local link: unresolved path
    '../../devdocs/libgit2/#Base.LibGit2.push_head!' in devdocs/libgit2.md
 !! Invalid local link: unresolved path
    '../../devdocs/libgit2/#Base.LibGit2.GitRevWalker' in devdocs/libgit2.md
 !! Invalid local link: unresolved path
    '../../devdocs/libgit2/#Base.LibGit2.GitHash' in devdocs/libgit2.md
 !! Invalid local link: unresolved path
    '../../devdocs/libgit2/#Base.LibGit2.push_head!' in devdocs/libgit2.md
```

to turn into:

```
 !! No doc found for reference '[`GitRevWalker`](@ref)'. [src/base/collections.md]
 !! No doc found for reference '[`GitHash`](@ref)'. [src/base/collections.md]
 !! No doc found for reference '[`push_head!`](@ref)'. [src/base/collections.md]
 !! No doc found for reference '[`GitRevWalker`](@ref)'. [src/base/collections.md]
 !! No doc found for reference '[`GitHash`](@ref)'. [src/base/collections.md]
 !! No doc found for reference '[`push_head!`](@ref)'. [src/base/collections.md]
```

which apparently fails the documentation build... I have no idea how to fix it but I temporarily turned off `strict` documentation build to see if the other tests pass.

**Comment by KristofferC** [comment]
Hmm:

```
Assertion failed!
Program: C:\projects\julia\usr\bin\julia.exe
File: /cygdrive/c/projects/julia/src/codegen.cpp, Line 1096
Expression: li->min_world <= world && li->max_world >= world
```

on AppVeyor.

More complete error:

```
markdown/Markdown.jl
docs/Docs.jl
C:\projects\julia\base\precompile.jl
This application has requested the Runtime to terminate it in an unusual way.
Please contact the application's support team for more information.
signal (22): SIGABRT
in expression starting at no file:0
raise at C:\windows\SYSTEM32\msvcrt.dll (unknown line)
abort at C:\windows\SYSTEM32\msvcrt.dll (unknown line)
_wassert at /usr/src/debug/mingw64-i686-runtime-5.0.3-1/misc\wassert.c:54
iob at C:\windows\SYSTEM32\msvcrt.dll (unknown line)
.rdata at C:\projects\julia\usr\bin\libjulia.dll (unknown line)
unknown function (ip: 091E7E1F)
unknown function (ip: 00CAE545)
unknown function (ip: 4679CB9F)
unknown function (ip: 0070005B)
unknown function (ip: 006F0071)
unknown function (ip: 00650069)
unknown function (ip: 00740062)
unknown function (ip: 005C0072)
unknown function (ip: 00750069)
unknown function (ip: 0069006B)
unknown function (ip: 005C0060)
unknown function (ip: 00730074)
unknown function (ip: 005C0071)
unknown function (ip: 00690061)
unknown function (ip: 005C006D)
unknown function (ip: 00750069)
unknown function (ip: 0069006B)
unknown function (ip: 002E0060)
unknown function (ip: 00780064)
Allocations: 379991728 (Pool: 379966701; Big: 25027); GC: 1815
Assertion failed!
Program: C:\projects\julia\usr\bin\julia.exe
File: /cygdrive/c/projects/julia/src/codegen.cpp, Line 1096
Expression: li->min_world <= world && li->max_world >= world
*** This error is usually fixed by running `make clean`. If the error persists, try `make cleanall`. ***
make[1]: *** [Makefile:221: /cygdrive/c/projects/julia/usr/lib/julia/sys.o] Error 1
make[1]: Leaving directory '/cygdrive/c/projects/julia'
make: *** [Makefile:101: julia-sysimg-release] Error 2
make: *** Waiting for unfinished jobs....
C:\projects\julia\base\precompile.jl
This application has requested the Runtime to terminate it in an unusual way.
```

File: base/deprecated.jl (hunk 3 above)
Comment (KristofferC) [reply base/deprecated.jl]: Hmm, the behavior right now is:

```
julia> Base.LibGit2.prompt("Hello?")
WARNING: Base.LibGit2 is deprecated, run `import LibGit2` instead.
 in module Base
┌ Warning: `LibGit2.prompt(msg::AbstractString; default::AbstractString="", password::Bool=false)` is deprecated, use `result = Base.prompt(msg, default=default, password=password); result === nothing ? "" : result` instead.
│   caller = top-level scope
└ @ Core :0
Hello?:yes
"yes"
```

Isn't that ok?

Comment (vchuravy) [reply base/deprecated.jl]: Oh I didn't think that would happen. Neat. You get a deprecation warning for `Base.LibGit2` and then a separate one for `prompt`. So as long as non of these were exported everything should be ok.

**Comment by KristofferC** [comment]
I'm going slightly crazy here...

When I disable the "LibGit2 moving to stdlib"-deprecation, FreeBSD tests pass but all other CI-systems fail because Documenter no longer works and they need documentation to run properly to even start the tests. This seems like a terrible idea in itself. Anyway, when I enable the LibGit2 deprecation, FreeBSD suddenly starts to fail. And everything passes locally. Wat.

**Comment by KristofferC** [comment]
CI straight flush. 🎉 

So what is left to do here is:

- [ ] Figure out what to do with `url(::Method)` when the `Method` is defined in a package:
https://github.com/JuliaLang/julia/blob/0f95988141373cebb968a2eed022d661b440293d/base/methodshow.jl#L215-L232. I suggest that we simply delete it since it assumed that packages are git repos which will not be the case.

- [ ] Figure out how to get documentation to pass with `strict = true` (see https://github.com/JuliaLang/julia/pull/25706#issuecomment-360169734). I have no idea about this, to be honest. These errors come when trying to reference LibGit2 docstrings from `Base.map` so one way would be to not have LibGit2 extend `Base.map` and instead define its own (similar to how it was done with `push!`).

Opinions?

**Comment by JeffBezanson** [comment]
I certainly wouldn't want to lose method URLs in jupyter; it's a cool feature.

**Comment by KristofferC** [comment]
Is https://github.com/JuliaLang/julia/pull/25706/commits/bbf2224f69941b4f6bf40f3bbacb397f3b2beb97 acceptable as a way out of this deadlock?

**Comment by JeffBezanson** [comment]
Sure, let's do that for now.

**Pull request merged**